### PR TITLE
Links: do not reposition container on selection change

### LINKED ISSUE #14872: Components: URL input popover animates repeatedly on text/selection change
Previously: #13617

**Describe the bug**

When changing the text within a linked segment, or moving the selection caret using arrow keys, the link popover animates repeatedly.

**To reproduce**

1. Navigate to Posts > Add New
2. Click the writing prompt
3. Insert some text
4. Apply a link to said text
5. Select within the linked fragment of text
6. Change the linked text, or simply use arrow keys within to change caret position
7. Note that the popover animates in response to each input

**Expected behavior**

The Popover should only animate once when first shown, not on input.

**Screenshots**

url-animation

**Desktop:**

 - OS: macOS Mojave 10.14.4 (18E226)
 - Browser: Chrome Version 73.0.3683.103 (Official Build) (64-bit)
 - Version: `master` (1df302121f0ab7df8611d01e28ad41db25f40d52)

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #14872. It also makes sure that the container is positioned relative to the link element rather than the selection/caret, so it doesn't move when you move selection inside a link.

There are several steps that came to the fix.

1. I wanted to modify `PositionedAtSelection` so that it accepts a `selector` prop, to select the closest matching element to calculate the position from rather than the caret.
2. In order to not recreate `PositionedAtSelection` for every change in selection, we need to use the format object reference as an indicator to recreate, as each DOM element is a different object instance.
3. Unfortunately, React does not accept object references as keys, so we have to derive a simpler state to use as the key. I used `getDerivedStateFromProps` to bump the key every time the active format object changes.
4. Something was wrong in setting the `activeFormats` in `onSelectionChange`. The formats need to be taken from the current value to maintain references, not from the created value (from DOM) on the selection change.
5. To ensure object references persist in the RichText component across selection changes and edits, I've taken out recreating the rich text value for every change. This actually results in other positive changes: no more `formatToValue` call for every input event, and no more need memoize it, and no more need for `updateFormats`, introduced in #14411.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by ellatrix** [comment]
The only remaining issue here is that the toolbar rerenders when you typing in the element. Fixing that now.

File: packages/format-library/src/link/inline.js [hunk 1]
```diff
@@ -222,7 +236,12 @@ class InlineLinkUI extends Component {
 
 		return (
 			<PositionedAtSelection
-				key={ `${ value.start }${ value.end }` /* Used to force rerender on selection change */ }
```
Comment (aduth) [inline packages/format-library/src/link/inline.js]: Possible dumb question: Couldn't a minimal solution just be to remove `key`?

Comment (ellatrix) [reply packages/format-library/src/link/inline.js]: What if you move the caret from one link to another?

Comment (aduth) [reply packages/format-library/src/link/inline.js]: > What if you move the caret from one link to another?

That's true. I wonder then if it could either act more like the `Autocomplete` component in [rendering `Popover` with a `getAnchorRect`](https://github.com/WordPress/gutenberg/blob/0b93ad47797a1ba2cea3c83a7f973d33abcd351f/packages/components/src/autocomplete/index.js#L469) that updates with the selection (but keeps the same element so not to unmount and re-trigger animation), or if `PositionedAtSelection` should update its styling in response to selection changes rather than [once at mount](https://github.com/WordPress/gutenberg/blob/0b93ad47797a1ba2cea3c83a7f973d33abcd351f/packages/components/src/positioned-at-selection/index.js#L50-L52).

I'm not totally sure the ramifications of either of these alternatives. Just trying to consider in terms of the least potential for impact.

File: packages/block-editor/src/components/rich-text/index.js [hunk 2]
```diff
@@ -459,12 +470,18 @@ export class RichText extends Component {
...
 			const { isCaretWithinFormattedText } = this.props;
-			const activeFormats = getActiveFormats( value );
+			const value = this.getRecord();
+			const newValue = {
+				...value,
+				start,
+				end,
+			};
+			delete newValue.activeFormats;
+			const activeFormats = getActiveFormats( newValue );
```
Comment (aduth) [inline packages/block-editor/src/components/rich-text/index.js]: If we're working with a shallow clone, seems we might as well assign `activeFormats` directly, rather than create another shallow clone below?

```js
const newValue = {
	...value,
	activeFormats: getActiveFormats( newValue ),
	start,
	end,
};
```

File: packages/block-editor/src/components/rich-text/index.js [hunk 3]
```diff
@@ -763,6 +781,7 @@ export class RichText extends Component {
 		// In all other cases, prevent default behaviour.
 		event.preventDefault();
 
+		const formats = createPrepareEditableTree( this.props )( value );
```
Comment (aduth) [inline packages/block-editor/src/components/rich-text/index.js]: I don't understand why this change is here.

Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js]: I'll comment, it's to make sure the boundaries work for annotations.

**Comment by ellatrix** [comment]
Updated the description:

> To ensure object references persist in the RichText component across selection changes and edits, I've taken out recreating the rich text value for every change. This actually results in other positive changes: no more formatToValue call for every input event, and no more need memoize it, and no more need for updateFormats, introduced in #14411.

**Comment by ellatrix** [comment]
While I think we eventually want all the changes here to RichText, I'm going to make an alternative PR that doesn't touch as much existing code as it does now.

**Comment by aduth** [comment]
Needs to be rebased after #14938

**Comment by aduth** [comment]
(I'm assuming this doesn't _need_ to land for Gutenberg 5.5 and have removed it from the milestone accordingly)

**Comment by ellatrix** [comment]
Closing for now.
